### PR TITLE
Allow custom steering functions

### PR DESCRIPTION
**Pull request by Hellisotherpeople**
introduced the following enhancements:

 Added a new parameter steering_method to the SteerElement dataclass and the add method. This allows users to provide a custom steering function that takes the tensor, coefficient, and try_keep_nr as input and returns the steering delta. If no custom steering function is provided, the default mean-based steering approach is used.

 Modified the _steer_vector_forward method to check if a custom steering function is provided for each steering element. If a custom function is provided, it is used to calculate the steering delta. Otherwise, the default mean-based approach is used.

With these changes, users can now define and use their own advanced steering techniques by providing a custom steering function. For example, they can implement weighted averaging, attention-based steering, or layer-wise steering by defining the appropriate function and passing it to the add method.

```python
def weighted_average_steering(tensor, coeff, try_keep_nr):
    weights = torch.softmax(tensor[:, try_keep_nr:, :], dim=1)
    weighted_tensor = weights * tensor[:, try_keep_nr:, :]
    delta = torch.sum(coeff * weighted_tensor, dim=1, keepdim=True)
    return delta

steer.add(
    layer_idx=12,
    coeff=0.8,
    text="example text",
    try_keep_nr=1,
    steering_method=weighted_average_steering,
)
```

In this example, the weighted_average_steering function calculates a weighted average of the tensor values based on softmax weights, and returns the steering delta. This custom steering function is then passed to the add method to apply the weighted average steering technique.

Further example functions we can pass in:

```python
def attention_based_steering(tensor, coeff, try_keep_nr):
    attention_weights = torch.softmax(torch.mean(tensor[:, try_keep_nr:, :], dim=-1), dim=-1)
    attention_weights = attention_weights.unsqueeze(-1).expand_as(tensor[:, try_keep_nr:, :])
    attended_tensor = attention_weights * tensor[:, try_keep_nr:, :]
    delta = torch.mean(coeff * attended_tensor, dim=1, keepdim=True)
    return delta
```
This steering function calculates attention weights based on the mean of the tensor values along the last dimension. The attention weights are then used to weight the tensor values, giving more importance to certain positions. The steering delta is calculated as the mean of the attended tensor values.

### CODE REVIEW

**Comment by Mihaiii** [comment]
Very nice and it makes a lot of sense!
I'll bump minor and release a new version in few minutes!

**Comment by Hellisotherpeople** [comment]
I notice that I pulled out the docstrings on my version, which was unintentional

**Comment by Mihaiii** [comment]
> I notice that I pulled out the docstrings on my version, which was unintentional

It's fine. In llm_steer's case, they made the code harder to read/follow.
But if/when you'd have some time, please consider updating the README.md with the added feature,
In any case, I made the release: https://pypi.org/project/llm-steer/1.1.0 . :)
